components/places/src/bookmark_sync/store.rs

```diff
@@ -0,0 +1,1233 @@
+                        "url": "https://www.example.com/"
+                    },
+                ]
+            }),
+        );
+
+        let store = BookmarksStore {
+            db: &conn,
+            client_info: &Cell::new(None),
+            local_time: Timestamp::now(),
+            remote_time: ServerTimestamp(0.0),
+        };
+        let tree = store.fetch_local_tree()?;
+
+        // should be each user root, plus the real root, plus the bookmark we added.
+        assert_eq!(
+            tree.guids().count(),
+            BookmarkRootGuid::user_roots().len() + 2
+        );
+
+        let node = tree
+            .node_for_guid(&"bookmark1___".into())
+            .expect("should exist");
+        assert_eq!(node.needs_merge, true);
+        assert_eq!(node.level(), 2);
+        assert_eq!(node.is_syncable(), true);
+
+        let node = tree
+            .node_for_guid(&BookmarkRootGuid::Unfiled.as_guid().into())
+            .expect("should exist");
+        assert_eq!(node.needs_merge, true);
+        assert_eq!(node.level(), 1);
+        assert_eq!(node.is_syncable(), true);
+
+        let node = tree
+            .node_for_guid(&BookmarkRootGuid::Menu.as_guid().into())
+            .expect("should exist");
+        assert_eq!(node.needs_merge, false);
+        assert_eq!(node.level(), 1);
+        assert_eq!(node.is_syncable(), true);
+
+        let node = tree
+            .node_for_guid(&BookmarkRootGuid::Root.as_guid().into())
+            .expect("should exist");
+        assert_eq!(node.needs_merge, false);
+        assert_eq!(node.level(), 0);
+        assert_eq!(node.is_syncable(), false);
+
+        // We should have changes.
+        assert_eq!(store.has_changes().unwrap(), true);
+        Ok(())
+    }
+
+    #[test]
+    fn test_apply_bookmark() {
+        assert_incoming_creates_local_tree(
+            json!([{
+                // A valid query (which actually looks just like a bookmark, but that's ok)
+                "id": "bookmark1___",
+                "type": "bookmark",
+                "parentid": "unfiled",
+                "parentName": "Unfiled Bookmarks",
+                "dateAdded": 1381542355843u64,
+                "title": "Some bookmark",
+                "bmkUri": "http://example.com",
+            },
+            {
+                "id": "unfiled",
+                "type": "folder",
+                "parentid": "root",
+                "dateAdded": 1381542355843u64,
+                "title": "Unfiled",
+                "children": ["bookmark1___"],
+
+            }]),
+            &BookmarkRootGuid::Unfiled.as_guid(),
+            json!({"children" : [{"guid": "bookmark1___", "url": "http://example.com"}]}),
+        );
+    }
+
+    #[test]
+    fn test_apply_query() {
+
+        // And finally, a more "functional" test - that our queries end up
+        // correctly in the local tree.
+        // XXX - this test fails for reasons I don't understand. If we write
```

Comment:
We talked about this on Slack, but, because we don't store `unfiled`, `query1______` is orphaned, so we'll delete it on the server. If you add a record for `unfiled` mentioning `query1______` in its children, it should apply.

Reply:
yeah, thanks - `test_apply_query` has this test reinstated.